zebra-chain/src/block/height.rs

```diff
@@ -83,18 +83,19 @@ impl Add<Height> for Height {
 }
 
 impl Sub<Height> for Height {
-    type Output = i32;
-
-    /// Panics if the inputs or result are outside the valid i32 range.
-    fn sub(self, rhs: Height) -> i32 {
-        // We construct heights from integers without any checks,
-        // so the inputs or result could be out of range.
-        let lhs = i32::try_from(self.0)
-            .expect("out of range input `self`: inputs should be valid Heights");
-        let rhs =
-            i32::try_from(rhs.0).expect("out of range input `rhs`: inputs should be valid Heights");
-        lhs.checked_sub(rhs)
-            .expect("out of range result: valid input heights should yield a valid result")
+    type Output = Option<Height>;
```

Comment:
Should this be `Option<u32>` instead?

`Height`s in Zebra represent the height of a block in the blockchain, but this operation returns the difference between two heights. (It's like `DateTime` and `Duration`, where a duration is the difference between two times.)

So the returned height might be numerically correct, but using it instead of the height of a block could lead to consensus rule mistakes in future changes.

Reply:
I think it depends on what the requirements are. If we don't have any code that uses negative `HeightDiff`s, then we could just return `None` for them. What do the existing callers of this code need?

We can always change it later if we need something more complicated.

I also think that some RPCs might manually construct heights that are above `Height::MAX`. We should check they aren't creating invalid heights.

We could also open a ticket to make the `u32` inside `Height` a private field, because until we do that, we can write code that constructs invalid heights, which can then cause panics later. But that's a huge change, 

Here's one possible minor change to the implementation:
```rust
pub type HeightDiff = i32;

// Delete impl Add<Height> for Height
// because it doesn't actually make sense to add two heights

// Do we want to return `None` rather than panicking, because other code can construct invalid heights?
// Do we want to return `None` rather than negative values, because they aren't needed?
// If we have two different invalid outcomes, do we want to return a `Result` instead?
impl Sub<Height> for Height {
    type Output = Option<HeightDiff>;
    ...
}

// This is the existing impl, but using the type alias
impl Add<HeightDiff> for Height {
    type Output = Option<Height>;
    ...
}

// This is the existing impl, but using the type alias
impl Sub<HeightDiff> for Height {
    type Output = Option<Height>;
    ...
}
```

We might also want to test anything that gets externally provided heights with inputs near `Height::MAX`:
* syncer
* inbound block downloader
* mempool
* transaction verifier
* RPCs